### PR TITLE
Denoise profile wavelets: change force by frequency and channel

### PR DESCRIPTION
**Pull request by rawfiner**
This pull request adds a GUI for the wavelets of denoise profile that is similar to equalizer.
The GUI is the same as in https://github.com/darktable-org/darktable/pull/1752

This does not change the underlying algorithm, but allows the user to finely control the denoising:
  - the user can change the force band by band. For instance, this is useful if user wants to preserve fine-grain noise
  - the user can change the force band by band for each channel. As in raw files there are less "red" and "blue" pixels than "green" pixels before demosaic, noise in red and blue channel has usually a coarser grain than in the green channel after demosaic, as the errors are "spread" by demosaic. The changes proposed here allow to take this into account in the denoising, but increasing the force for red and blue without changing the green channel.

I choosed to use a gtk stack for the GUI to solve the issue we had before:
When duplicating an instance that is in wavelet mode, the "patch size" slider was appearing.
bug_denoise_profile
With a stack, this is fixed.

How to get better denoising with this change?
There are plenty of ways:
- You can use the "all" curve to preserve the fine details more in case of coarse grain noise, but lowering the curve on the right.
- You can use the gray channel of channel mixer module to see only R, G or B channel, and then tune the denoising channel per channel. Then disable the channel mixer to see the result, and add another instance in color blend mode if needed.
- Sometimes in color blend mode, some small green pixels are not corrected, which forces to push the force slider a lot. With this change, the fix is easy: take the green curve and push it toward the top on finest scale:
Before (please look at 100% zoom level, as we are talking about pixel-level changes here):
capture du 2018-10-16 13-39-48
After:
capture du 2018-10-16 13-40-03

Finally, in order to control easily the force with the curve, you should set the force slider around 0.3, which gives much more flexibility than 1.

### CODE REVIEW

**Comment by TurboGit** [comment]
Do you confirm that without touching the GUI to adjust all or R, G, B channels individually the actual result is exactly the same as before? That is, the wavelet strength is working as in the previous implementation. I'm asking, because a very important point here is to keep a simple usage for this module while allowing some more tweaking on the channels.

**Comment by TurboGit** [comment]
Forgot to say, I've tested that on a quite noisy picture and I'm impressed by the results!

**Comment by rawfiner** [comment]
> Do you confirm that without touching the GUI to adjust all or R, G, B channels individually the actual result is exactly the same as before? That is, the wavelet strength is working as in the previous implementation. I'm asking, because a very important point here is to keep a simple usage for this module while allowing some more tweaking on the channels.

Yes. See this part of the code:

`
    float adjt[3] = { 8.0f, 8.0f, 8.0f };
    int offset_scale = DT_IOP_DENOISE_PROFILE_BANDS - max_scale;
    // current scale number is s+offset_scale
    // for instance, largest scale is DT_IOP_DENOISE_PROFILE_BANDS
    // max_scale only indicates the number of scales to process at THIS
    // zoom level, it does NOT corresponds to the the maximum number of scales.
    // in other words, max_scale is the maximum number of VISIBLE scales.
    // That is why we have this "s+offset_scale"
    float band_force_exp_2 = d->force[denoiseprofile_all][DT_IOP_DENOISE_PROFILE_BANDS - (s + offset_scale + 1)];
    band_force_exp_2 *= band_force_exp_2;
    band_force_exp_2 *= 4; // scale to [0,4]. 1 is the neutral curve point
    for(int ch = 0; ch < 3; ch++)
    {
      adjt[ch] *= band_force_exp_2;
    }
    band_force_exp_2 = d->force[denoiseprofile_R][DT_IOP_DENOISE_PROFILE_BANDS - (s + offset_scale + 1)];
    band_force_exp_2 *= band_force_exp_2;
    band_force_exp_2 *= 4; // scale to [0,4]. 1 is the neutral curve point
    adjt[0] *= band_force_exp_2;
    band_force_exp_2 = d->force[denoiseprofile_G][DT_IOP_DENOISE_PROFILE_BANDS - (s + offset_scale + 1)];
    band_force_exp_2 *= band_force_exp_2;
    band_force_exp_2 *= 4; // scale to [0,4]. 1 is the neutral curve point
    adjt[1] *= band_force_exp_2;
    band_force_exp_2 = d->force[denoiseprofile_B][DT_IOP_DENOISE_PROFILE_BANDS - (s + offset_scale + 1)];
    band_force_exp_2 *= band_force_exp_2;
    band_force_exp_2 *= 4; // scale to [0,4]. 1 is the neutral curve point
    adjt[2] *= band_force_exp_2; `

Basically, before, adjt was always equal to 8.0f. In the new code, adjt has 3 values, one per channel.
We take the point value in the curve, which is 0.5 is curve is not moved, square that value, which gives 0.25, and multiply by 4, which gives 1.
adjt elements are multiplied by this value, so multiplied by 1 if the curve is flat.
The for loop multiplies the adjt elements of all channels by the value computed from the "all" curve, while the code which is after the for loop multiply the adjt elements channel by channel, considering the R, G, and B curve

**Comment by rawfiner** [comment]
Sorry for the code formatting in the quotes, I am not sure yet how to refer to code from https://github.com/darktable-org/darktable/pull/1753/files

**Comment by aurelienpierre** [comment]
@rawfiner you can use 3 ` to enclose the code blocks.

**Comment by TurboGit** [comment]
I've tested it, this is really nice. But please if you read this test it before it is merged. I'd like to have this in 2.6, so more testing would be nice.

**Comment by aurelienpierre** [comment]
It crashes when using old presets and usually doesn't display the existing profile when available (the combobox is empty) in wavelet and non-local means :
```
(gdb) bt full
#0  0x00007ffff7963415 in catmull_rom_val (n=7, x=<optimized out>, xval=<optimized out>, y=0x7fffffffd4c0, tangents=0x0) at /home/aurelien/Documents/Programmes/darktable/src/common/curve_tools.c:567
        ival = <optimized out>
        m0 = Python Exception <class 'gdb.MemoryError'> Cannot access memory at address 0x14: 
#1  0x00007fff80426ef1 in dt_draw_curve_calc_value (x=0.000273924117, c=0x555559258b80) at /home/aurelien/Documents/Programmes/darktable/src/gui/draw.h:214
        xa = {-1, 0, 0, 0, 0, 0, 1, 0, 1.99384797e+13, 3.0611365e-41, 0, 0, 1.99384797e+13, 3.0611365e-41, -1.74169807e+32, 4.59163468e-41, 2.04745789e+13, 3.0611365e-41, -1.76523422e+32, 4.59163468e-41}
        ya = {0, 0, 0, 0, 0, 0, 0, -3.8272803e-31, -nan(0x7fd5d0), 4.59163468e-41, 0, 0, 0, 0, 4.35041532e+14, 3.0611365e-41, -1.0507078e+33, 4.59163468e-41, 2.80259693e-45, 0}
        val = <optimized out>
        ypp = 0x0
        inset = <optimized out>
        allocation = {x = 0, y = 32, width = 303, height = 227}
        height = 215
        width = <optimized out>
        self = <optimized out>
        c = <optimized out>
        ch = <optimized out>
#2  0x00007fff80426ef1 in denoiseprofile_button_press (widget=<optimized out>, event=0x7fff4c0030c0, user_data=<optimized out>) at /home/aurelien/Documents/Programmes/darktable/src/iop/denoiseprofile.c:2498
        inset = <optimized out>
        allocation = {x = 0, y = 32, width = 303, height = 227}
        height = 215
        width = <optimized out>
        self = <optimized out>
        c = <optimized out>
        ch = <optimized out>
#3  0x00007ffff653d15b in  () at /usr/lib/x86_64-linux-gnu/libgtk-3.so.0
#4  0x00007ffff509810d in g_closure_invoke () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#5  0x00007ffff50aafce in  () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#6  0x00007ffff50b301f in g_signal_emit_valist () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#7  0x00007ffff50b409f in g_signal_emit () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#8  0x00007ffff6684b04 in  () at /usr/lib/x86_64-linux-gnu/libgtk-3.so.0
#9  0x00007ffff653a1ce in  () at /usr/lib/x86_64-linux-gnu/libgtk-3.so.0
#10 0x00007ffff653c2a8 in gtk_main_do_event () at /usr/lib/x86_64-linux-gnu/libgtk-3.so.0
#11 0x00007ffff604d765 in  () at /usr/lib/x86_64-linux-gnu/libgdk-3.so.0
#12 0x00007ffff607df82 in  () at /usr/lib/x86_64-linux-gnu/libgdk-3.so.0
#13 0x00007ffff6c5f287 in g_main_context_dispatch () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#14 0x00007ffff6c5f4c0 in  () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#15 0x00007ffff6c5f7d2 in g_main_loop_run () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#16 0x00007ffff653b385 in gtk_main () at /usr/lib/x86_64-linux-gnu/libgtk-3.so.0
#17 0x00007ffff7a5518a in dt_gui_gtk_run (gui=<optimized out>) at /home/aurelien/Documents/Programmes/darktable/src/gui/gtk.c:1242
        widget = <optimized out>
        allocation = {x = 369, y = 19, width = 375, height = 575}
#18 0x0000555555554830 in main (argc=<optimized out>, argv=<optimized out>) at /home/aurelien/Documents/Programmes/darktable/src/main.c:83
```

**Comment by jpg54** [comment]
I have been using the new RawFiner modifications in Denoise profile wavelets for some time. I appreciate that he was able to include in darktable's sources the research and experiments (and the results he drew from them) that he did with the dtQStyles he shared.
I would like to see them in the next version 2.6.

File: src/iop/denoiseprofile.c [hunk 1]
```diff
@@ -43,22 +44,36 @@
...
+#define DT_IOP_DENOISE_PROFILE_BANDS 5
+
 typedef enum dt_iop_denoiseprofile_mode_t
 {
   MODE_NLMEANS = 0,
   MODE_WAVELETS = 1
 } dt_iop_denoiseprofile_mode_t;
 
+typedef enum dt_iop_denoiseprofile_channel_t {
+  denoiseprofile_all = 0,
```
Comment (TurboGit) [inline src/iop/denoiseprofile.c]: All entries should be capitalize:

DENOISEPROFILE_ALL = 0,
...

File: src/iop/denoiseprofile.c [hunk 2]
```diff
@@ -2085,20 +2227,402 @@ void gui_update(dt_iop_module_t *self)
...
+                           gtk_label_new(_("G")));
+  gtk_notebook_append_page(GTK_NOTEBOOK(g->channel_tabs), GTK_WIDGET(gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 0)),
+                           gtk_label_new(_("B")));
+
+  gtk_widget_show_all(GTK_WIDGET(gtk_notebook_get_nth_page(g->channel_tabs, g->channel)));
+  gtk_notebook_set_current_page(GTK_NOTEBOOK(g->channel_tabs), g->channel);
+  g_signal_connect(G_OBJECT(g->channel_tabs), "switch_page", G_CALLBACK(denoiseprofile_tab_switch), self);
+
+  g->channel = dt_conf_get_int("plugins/darkroom/denoiseprofile/gui_channel");
+  int ch = (int)g->channel;
```
Comment (TurboGit) [inline src/iop/denoiseprofile.c]: could probably be : const int ch = ...

File: src/iop/denoiseprofile.c [hunk 3]
```diff
@@ -94,20 +133,22 @@ static dt_noiseprofile_t dt_iop_denoiseprofile_get_auto_profile(dt_iop_module_t
 int legacy_params(dt_iop_module_t *self, const void *const old_params, const int old_version,
                   void *new_params, const int new_version)
 {
-  if((old_version == 1 || old_version == 2) && new_version == 3)
+  if((old_version == 1 || old_version == 2 || old_version == 3) && new_version == 4)
   {
     dt_iop_denoiseprofile_params_t *o = (dt_iop_denoiseprofile_params_t *)old_params;
```
[the remark this answers is not in the document]
Comment (TurboGit) [reply src/iop/denoiseprofile.c]: See exposure.c for inspiration for example.

File: src/iop/denoiseprofile.c [hunk 4]
```diff
@@ -94,20 +133,22 @@ static dt_noiseprofile_t dt_iop_denoiseprofile_get_auto_profile(dt_iop_module_t
...
     if(old_version == 1)
     {
-      // one more parameter and changed parameters in case we autodetected a profile
-      memcpy(n, o, sizeof(dt_iop_denoiseprofile_params_t) - sizeof(uint32_t));
       n->mode = MODE_NLMEANS;
     }
     else
     {
-      memcpy(n, o, sizeof(dt_iop_denoiseprofile_params_t));
+      n->mode = o->mode;
```
[the remark this answers is not in the document]
Comment (rawfiner) [reply src/iop/denoiseprofile.c]: As long as the struct new members are added at the end of the struct, one can continue to use the same type, as long as no access to new elements are made. But I agree that code, even if correct, can be harder to understand.

**Comment by rawfiner** [comment]
Thanks a lot for the code review.

Concerning legacy_param, I had made a mistake: the curve parameters were sometimes not initialized. That is why it was crashing on presets.

I created a new type for the old set of param, but used it for all older versions, as it was done like this before this pull request.
As the new members in the struct have always been added at the end, the type defined by the struct can be used on old params as long as no access to new members are made.
However, I can understand that this results in code harder to read (and more risky to modify), so tell me if you want me to change this and define different types for all 4 versions.

**Comment by rawfiner** [comment]
@aurelienpierre, could you tell me in which cases the combobox is empty?
The crash with presets should be fixed now.

**Comment by TurboGit** [comment]
All good now thanks.

**Comment by TurboGit** [comment]
All good now for a wider audience! Thanks.
